Lib/ufo2ft/kernFeatureWriter.py

```diff
@@ -192,20 +238,73 @@ def _addGlyphClasses(self, lines):
         for key, members in sorted(self.groups.items()):
             lines.append("%s = [%s];" % (key, " ".join(members)))
 
-    def _addKerning(self, lines, kerning, enum=False, ignoreZero=False):
+    def _splitRtlKerning(self):
+        """Split RTL kerning into separate dictionaries."""
+
+        self.rtlGlyphPairKerning = {}
+        self.rtlLeftClassKerning = {}
+        self.rtlRightClassKerning = {}
+        self.rtlClassPairKerning = {}
+
+        classes = self._getClasses()
+        allKerning = (
+            (self.glyphPairKerning, self.rtlGlyphPairKerning, (False, False)),
+            (self.leftClassKerning, self.rtlLeftClassKerning, (True, False)),
+            (self.rightClassKerning, self.rtlRightClassKerning, (False, True)),
+            (self.classPairKerning, self.rtlClassPairKerning, (True, True)))
+
+        for origKerning, rtlKerning, (leftIsClass, rightIsClass) in allKerning:
+            for pair in list(origKerning.keys()):
+                lhs, rhs = pair
+                leftGlyphs = classes[lhs] if leftIsClass else [lhs]
+                rightGlyphs = classes[rhs] if rightIsClass else [rhs]
+                if any(self._glyphIsRtl(g) for g in leftGlyphs + rightGlyphs):
+                    rtlKerning[pair] = origKerning.pop(pair)
+
+    def _addKerning(self, lines, kerning, rtl=False, enum=False,
+                    ignoreZero=False):
         """Add kerning rules for a mapping of pairs to values."""
 
         enum = "enum " if enum else ""
+        valstr = "<%(val)d 0 %(val)d 0>" if rtl else "%(val)d"
+        lineFormat = "    %spos %%(lhs)s %%(rhs)s %s;" % (enum, valstr)
         for (left, right), val in sorted(kerning.items()):
             if val == 0 and ignoreZero:
                 continue
-            lines.append("    %spos %s %s %d;" % (enum, left, right, val))
+            lines.append(lineFormat % {'lhs': left, 'rhs': right, 'val': val})
+
+    def _addLookupReferences(self, lines, languageSystems, lookupName):
+        """Add references to lookup for a set of language systems.
+
+        Language systems are passed in as a dictionary mapping scripts to lists
+        of languages.
+        """
+
+        for script, langs in sorted(languageSystems.items()):
+            lines.append("script %s;" % script)
+            for lang in sorted(langs):
```

Comment:
you shouldn't sort languages alphabetically here.
makeotf complains that 'dflt' must precede language-specific behavior.
```
makeotfexe [FATAL] <Ubuntu-Bold> dflt must precede language-specific behavior (text was "dflt") [tmpbVoiP4 8966]
``
When 'languagesystems' are defined in a feature file, they are usually already sorted in the correct order by the font developer, so you could simply use an OrderedDict and keep the original sorting of scripts and languages within each scripts.